File: app/src/main/java/dev/ragnarok/fenrir/fragment/VideoAlbumsByVideoFragment.java
package dev.ragnarok.fenrir.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.ActivityFeatures;
import dev.ragnarok.fenrir.activity.ActivityUtils;
import dev.ragnarok.fenrir.adapter.VideoAlbumsNewAdapter;
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment;
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener;
import dev.ragnarok.fenrir.model.VideoAlbum;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.VideoAlbumsByVideoPresenter;
import dev.ragnarok.fenrir.mvp.view.IVideoAlbumsByVideoView;
import dev.ragnarok.fenrir.place.PlaceFactory;
import dev.ragnarok.fenrir.util.Objects;
import dev.ragnarok.fenrir.util.ViewUtils;

public class VideoAlbumsByVideoFragment extends BaseMvpFragment<VideoAlbumsByVideoPresenter, IVideoAlbumsByVideoView>
        implements VideoAlbumsNewAdapter.Listener, IVideoAlbumsByVideoView {

    private SwipeRefreshLayout mSwipeRefreshLayout;
    private VideoAlbumsNewAdapter mAdapter;
    private TextView mEmpty;

    public static VideoAlbumsByVideoFragment newInstance(Bundle args) {
        VideoAlbumsByVideoFragment fragment = new VideoAlbumsByVideoFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public static VideoAlbumsByVideoFragment newInstance(int accountId, int ownerId, int video_ownerId, int video_Id) {
        return newInstance(buildArgs(accountId, ownerId, video_ownerId, video_Id));
    }

    public static Bundle buildArgs(int accountId, int ownerId, int video_ownerId, int video_Id) {
        Bundle args = new Bundle();
        args.putInt(Extra.ACCOUNT_ID, accountId);
        args.putInt(Extra.OWNER_ID, ownerId);
        args.putInt(Extra.OWNER, video_ownerId);
        args.putInt(Extra.VIDEO, video_Id);
        return args;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_video_albums_by_video, container, false);
        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));
        RecyclerView recyclerView = root.findViewById(R.id.recycler_view);

        mSwipeRefreshLayout = root.findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(() -> callPresenter(VideoAlbumsByVideoPresenter::fireRefresh));

        ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout);

        mEmpty = root.findViewById(R.id.empty);

        int columns = requireActivity().getResources().getInteger(R.integer.videos_column_count);
        StaggeredGridLayoutManager manager = new StaggeredGridLayoutManager(columns, StaggeredGridLayoutManager.VERTICAL);

        recyclerView.setLayoutManager(manager);
        recyclerView.addOnScrollListener(new PicassoPauseOnScrollListener(VideoAlbumsNewAdapter.PICASSO_TAG));

        mAdapter = new VideoAlbumsNewAdapter(requireActivity(), Collections.emptyList());
        mAdapter.setListener(this);
        recyclerView.setAdapter(mAdapter);

        resolveEmptyTextVisibility();
        return root;
    }

    @Override
    public void onClick(VideoAlbum album) {
        callPresenter(p -> p.fireItemClick(album));
    }

    @Override
    public void displayData(@NonNull List<VideoAlbum> data) {
        if (Objects.nonNull(mAdapter)) {
            mAdapter.setData(data);
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void notifyDataAdded(int position, int count) {
        if (Objects.nonNull(mAdapter)) {
            mAdapter.notifyItemRangeInserted(position, count);
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void displayLoading(boolean loading) {
        if (Objects.nonNull(mSwipeRefreshLayout)) {
            mSwipeRefreshLayout.post(() -> mSwipeRefreshLayout.setRefreshing(loading));
        }
    }

    private void resolveEmptyTextVisibility() {
        if (Objects.nonNull(mEmpty) && Objects.nonNull(mAdapter)) {
            mEmpty.setVisibility(mAdapter.getItemCount() == 0 ? View.VISIBLE : View.GONE);
        }
    }

    @Override
    public void openAlbum(int accountId, int ownerId, int albumId, String action, String title) {
        PlaceFactory.getVideoAlbumPlace(accountId, ownerId, albumId, action, title).tryOpenWith(requireActivity());
    }

    @Override
    public void notifyDataSetChanged() {
        if (Objects.nonNull(mAdapter)) {
            mAdapter.notifyDataSetChanged();
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        ActionBar actionBar = ActivityUtils.supportToolbarFor(this);
        if (actionBar != null) {
            actionBar.setTitle(R.string.videos_albums);
            actionBar.setSubtitle(null);
        }
        new ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(false)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity());
    }

    @NonNull
    @Override
    public IPresenterFactory<VideoAlbumsByVideoPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        return () -> {
            int ownerId1 = requireArguments().getInt(Extra.OWNER_ID);
            int accountId = requireArguments().getInt(Extra.ACCOUNT_ID);
            int owner = requireArguments().getInt(Extra.OWNER);
            int video = requireArguments().getInt(Extra.VIDEO);
            return new VideoAlbumsByVideoPresenter(accountId, ownerId1, owner, video, saveInstanceState);
        };
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/videoalbumsbyvideo/VideoAlbumsByVideoFragment.kt
package dev.ragnarok.fenrir.fragment.videoalbumsbyvideo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.ActivityFeatures
import dev.ragnarok.fenrir.activity.ActivityUtils.supportToolbarFor
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment
import dev.ragnarok.fenrir.fragment.base.core.IPresenterFactory
import dev.ragnarok.fenrir.fragment.videoalbums.VideoAlbumsNewAdapter
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener
import dev.ragnarok.fenrir.model.VideoAlbum
import dev.ragnarok.fenrir.place.PlaceFactory.getVideoAlbumPlace
import dev.ragnarok.fenrir.util.ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme

class VideoAlbumsByVideoFragment :
    BaseMvpFragment<VideoAlbumsByVideoPresenter, IVideoAlbumsByVideoView>(),
    VideoAlbumsNewAdapter.Listener, IVideoAlbumsByVideoView {
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    private var mAdapter: VideoAlbumsNewAdapter? = null
    private var mEmpty: TextView? = null
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_video_albums_by_video, container, false)
        (requireActivity() as AppCompatActivity).setSupportActionBar(root.findViewById(R.id.toolbar))
        val recyclerView: RecyclerView = root.findViewById(R.id.recycler_view)
        mSwipeRefreshLayout = root.findViewById(R.id.refresh)
        mSwipeRefreshLayout?.setOnRefreshListener {
            presenter?.fireRefresh()
        }
        setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout)
        mEmpty = root.findViewById(R.id.empty)
        val columns = requireActivity().resources.getInteger(R.integer.videos_column_count)
        val manager = StaggeredGridLayoutManager(columns, StaggeredGridLayoutManager.VERTICAL)
        recyclerView.layoutManager = manager
        recyclerView.addOnScrollListener(PicassoPauseOnScrollListener(VideoAlbumsNewAdapter.PICASSO_TAG))
        mAdapter = VideoAlbumsNewAdapter(requireActivity(), emptyList())
        mAdapter?.setListener(this)
        recyclerView.adapter = mAdapter
        resolveEmptyTextVisibility()
        return root
    }

    override fun onClick(album: VideoAlbum) {
        presenter?.fireItemClick(
            album
        )
    }

    override fun displayData(data: List<VideoAlbum>) {
        if (mAdapter != null) {
            mAdapter?.setData(data)
            resolveEmptyTextVisibility()
        }
    }

    override fun notifyDataAdded(position: Int, count: Int) {
        if (mAdapter != null) {
            mAdapter?.notifyItemRangeInserted(position, count)
            resolveEmptyTextVisibility()
        }
    }

    override fun displayLoading(loading: Boolean) {
        mSwipeRefreshLayout?.post { mSwipeRefreshLayout?.isRefreshing = loading }
    }

    private fun resolveEmptyTextVisibility() {
        if (mEmpty != null && mAdapter != null) {
            mEmpty?.visibility =
                if (mAdapter?.itemCount == 0) View.VISIBLE else View.GONE
        }
    }

    override fun openAlbum(
        accountId: Int,
        ownerId: Int,
        albumId: Int,
        action: String?,
        title: String?
    ) {
        getVideoAlbumPlace(
            accountId,
            ownerId,
            albumId,
            action,
            title
        ).tryOpenWith(requireActivity())
    }

    override fun notifyDataSetChanged() {
        if (mAdapter != null) {
            mAdapter?.notifyDataSetChanged()
            resolveEmptyTextVisibility()
        }
    }

    override fun onResume() {
        super.onResume()
        val actionBar = supportToolbarFor(this)
        if (actionBar != null) {
            actionBar.setTitle(R.string.videos_albums)
            actionBar.subtitle = null
        }
        ActivityFeatures.Builder()
            .begin()
            .setHideNavigationMenu(false)
            .setBarsColored(requireActivity(), true)
            .build()
            .apply(requireActivity())
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<VideoAlbumsByVideoPresenter> {
        return object : IPresenterFactory<VideoAlbumsByVideoPresenter> {
            override fun create(): VideoAlbumsByVideoPresenter {
                val ownerId1 = requireArguments().getInt(Extra.OWNER_ID)
                val accountId = requireArguments().getInt(Extra.ACCOUNT_ID)
                val owner = requireArguments().getInt(Extra.OWNER)
                val video = requireArguments().getInt(Extra.VIDEO)
                return VideoAlbumsByVideoPresenter(
                    accountId,
                    ownerId1,
                    owner,
                    video,
                    saveInstanceState
                )
            }
        }
    }

    companion object {
        fun newInstance(args: Bundle?): VideoAlbumsByVideoFragment {
            val fragment = VideoAlbumsByVideoFragment()
            fragment.arguments = args
            return fragment
        }

        fun newInstance(
            accountId: Int,
            ownerId: Int,
            video_ownerId: Int,
            video_Id: Int
        ): VideoAlbumsByVideoFragment {
            return newInstance(buildArgs(accountId, ownerId, video_ownerId, video_Id))
        }

        fun buildArgs(accountId: Int, ownerId: Int, video_ownerId: Int, video_Id: Int): Bundle {
            val args = Bundle()
            args.putInt(Extra.ACCOUNT_ID, accountId)
            args.putInt(Extra.OWNER_ID, ownerId)
            args.putInt(Extra.OWNER, video_ownerId)
            args.putInt(Extra.VIDEO, video_Id)
            return args
        }
    }
}